File: isItPoem.rb
def rhyme? a
  return a.collect { |t| t[1] }.map { |n| n[-3..-1] }.uniq.length == 1
end


def isItPoem r, t
  t.each_slice(r.length) { |n|
    tmp = r.zip(n)
    r.uniq.each { |p| return false if !rhyme? tmp.select { |t| t[0] == p } }
  }
  true
end

File: codefights/lib/is_it_poem.ex
defmodule IsItPoem do
  @moduledoc false

  @spec isItPoem([Integer, ...], [String.t, ...]) :: boolean
  def isItPoem(rhyme, text) do
    Enum.map(Enum.chunk(text, length rhyme), fn chunk ->
      rhyme_zip = Enum.zip rhyme, chunk
      Enum.uniq(rhyme)
      |> Enum.map(fn num ->
        Enum.filter(rhyme_zip, fn {num2, _} -> num == num2 end)
        |> Enum.map(fn {_, text} -> String.slice(text, -3..-1) end)
        |> Enum.uniq
        |> length
      end)
    end)
    |> List.flatten
    |> Enum.all?(&(&1 == 1))
  end
end
